sycl/include/CL/sycl/access/access.hpp

```diff
@@ -104,6 +115,16 @@ struct PtrValueType<ElementType, access::address_space::global_space> {
   using type = __OPENCL_GLOBAL_AS__ ElementType;
 };
 
+template <typename ElementType>
+struct PtrValueType<ElementType, access::address_space::device_space> {
+  using type = __OPENCL_GLOBAL_DEVICE_AS__ ElementType;
```

Comment:
This is likely to not be portable and cause ICE rather than a clear error. As this overlaps with `global`, the address space should fallback to `__OPENCL_GLOBAL_AS__` if the backend does not handles this address space.

Reply:
As DPCPP compiler generated SPIR-V code - this mechanism is currently moved to the SPIR-V translator (basically during reversed translation from SPIR-V to LLVM IR there is an option added - without this option passed the translator will generated global address space instead of global_device / global_host address space. So if someone would like to support these address spaces in their backend - it's needed to add this option in the backend's driver.

Reply:
> As DPCPP compiler generated SPIR-V code

or PTX via the NVPTX backend without going through SPIR-V

Reply:
Got it, thanks!

Reply:
Actually, not, I don't really get it. What target is used for NVPTX? I mean, that in clang part of the feature we have added definitions for these new address spaces like this:
`--- a/clang/lib/Basic/Targets/NVPTX.h
+++ b/clang/lib/Basic/Targets/NVPTX.h
@@ -30,6 +30,8 @@ static const unsigned NVPTXAddrSpaceMap[] = {
     0, // opencl_private
     // FIXME: generic has to be added to the target
     0, // opencl_generic
+    1, // opencl_global_device
+    1, // opencl_global_host
       1, // cuda_device
`

If for NVPTX we compile with spir-unknown-unknown triple, than the code above is indeed a problem. But if not - I don't see any issues.

Reply:
> I don't really get it. What target is used for NVPTX? 

`nvptx64-nvidia-cuda-sycldevice`

> I don't see any issues

The issue is in the mangler, given the current definition of the address space mapping

```cpp
void foo(global_ptr<int>::pointer_t p) { [...] }
void foo(device_ptr<int>::pointer_t p) { [...] }
```
This will cause the compiler to mangle the 2 `foo` overloads in the same way.

There is 2 solutions to it:
 - Having a new mangling scheme, but I'm not sure how it should be done (@bader ping for this)
 - a SYCL solution: make the address space available if and only if the target actually supports it.

Note: this is kind of a corner case for now, I pointing this out so you are aware of it. I'm more concerned about the naming here.

Reply:
Thanks you very much for your feedback. I'll think about these options.

Reply:
I would like to leave this corner case unresolved for now. One of the possible solutions is to expand authority of `sycl_enable_usm_address_spaces` option added in https://github.com/intel/llvm/pull/1986 , but in this case this option (which originally was considered as a temporary solution) will stay in the compiler.

Reply:
@MrSidims, could you open a GitHub issue to track/discuss solution this problem, please?